src/common/logging.rs

```diff
@@ -0,0 +1,272 @@
+pub fn stage2_init_pre_pivot_root_tmpfs_log_handler() {
+    let stage1_logfile = get_stage_tmpfs_logfile_path(Stage::S1);
+    let stage2_init_logfile = get_stage_tmpfs_logfile_path(Stage::S2Init);
+
+    // copy files to data partition
+    copy_tmpfs_log_to_data_partition(stage1_logfile.as_str(), LOG_PRE_UNMOUNT_DATA_PART_DEST);
+    copy_tmpfs_log_to_data_partition(stage2_init_logfile.as_str(), LOG_PRE_UNMOUNT_DATA_PART_DEST);
+}
+
+// Error handling if an error occurs in stage2-init prior to calling pivot_root
+pub fn stage2_init_pre_pivot_root_err_handler(log_to_balenaos: bool) -> ! {
+    if log_to_balenaos {
+        stage2_init_pre_pivot_root_tmpfs_log_handler();
+    }
+
+    reboot();
+}
+
+// Helper function to dump logs
+// Caters for state after calling pivot_root
+pub fn stage2_init_post_pivot_root_tmpfs_log_handler() {
+    // if pivot_root called, stage1 log will be relative to old root mountpoint
+    let stage1_logfile = format!(
+        "{}/{}",
+        OLD_ROOT_MP,
+        get_stage_tmpfs_logfile_path(Stage::S1)
+    );
+
+    let stage2_init_logfile = get_stage_tmpfs_logfile_path(Stage::S2Init);
+
+    // data part will be relative to old mount point
+    let destination = format!("{}/{}", OLD_ROOT_MP, LOG_PRE_UNMOUNT_DATA_PART_DEST);
+
+    // copy files to data partition
+    copy_tmpfs_log_to_data_partition(stage1_logfile.as_str(), destination.as_str());
+    copy_tmpfs_log_to_data_partition(stage2_init_logfile.as_str(), destination.as_str());
+}
+
+// Error handling if an error occurs in stage2-init after calling pivot_root
+pub fn stage2_init_post_pivot_root_err_handler() -> ! {
+    stage2_init_post_pivot_root_tmpfs_log_handler();
+    reboot();
+}
+
+// Helper function to dump logs
+// Caters for state in stage2 worker prior to unmounting partitions
+pub fn stage2_pre_unmount_tmpfs_log_handler() {
+    let stage1_logfile = format!(
+        "{}/{}",
+        OLD_ROOT_MP,
+        get_stage_tmpfs_logfile_path(Stage::S1)
+    );
+    let stage2_init_logfile = get_stage_tmpfs_logfile_path(Stage::S2Init);
+    let stage2_logfile = get_stage_tmpfs_logfile_path(Stage::Stage2);
+
+    // copy files to data partition
+    let destination = format!("{}/{}", OLD_ROOT_MP, LOG_PRE_UNMOUNT_DATA_PART_DEST);
+
+    copy_tmpfs_log_to_data_partition(stage1_logfile.as_str(), destination.as_str());
+    copy_tmpfs_log_to_data_partition(stage2_init_logfile.as_str(), destination.as_str());
+    copy_tmpfs_log_to_data_partition(stage2_logfile.as_str(), destination.as_str());
+}
+// Error handling if an error occurs in stage2 worker process before unmounting partitions
+pub fn stage2_pre_unmount_err_handler(log_to_balenaos: bool) -> ! {
+    if log_to_balenaos {
+        stage2_pre_unmount_tmpfs_log_handler();
+    }
+
+    reboot();
+}
+
+// Helper function to dump logs
+// Caters for state in stage2 worker after unmounting partitions
+pub fn stage2_post_unmount_tmpfs_log_handler(s2_config: &Stage2Config) -> Result<()> {
```

Comment:
Use a more descriptive name, like `persist_stage2_tmpfs_log()`. It will make the calling code easier to understand what it does.